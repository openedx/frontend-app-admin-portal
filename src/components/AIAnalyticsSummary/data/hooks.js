import { useEffect, useState } from 'react';
import { logError } from '@edx/frontend-platform/logging';
import LmsApiService from '../../../data/services/LmsApiService';

const useAIAnalyticsSummary = (enterpriseId, insights) => {
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);
  const [analyticsSummaryData, setAnalyticsSummaryData] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await LmsApiService.generateAIAnalyticsSummary(enterpriseId, insights);
        setAnalyticsSummaryData(response.data);
      } catch (err) {
        setError(err);
        logError(err);
      } finally {
        setIsLoading(false);
      }
    };

    if (enterpriseId && insights) {
      fetchData();
    } else {
      setIsLoading(false);
    }
  }, [enterpriseId, insights]);

  return {
    isLoading,
    error,
    data: analyticsSummaryData,
  };
};

export default useAIAnalyticsSummary;
