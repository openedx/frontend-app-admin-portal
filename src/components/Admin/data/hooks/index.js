export { default as useGenericTableData } from './useGenericTableData';
export { default as useModuleActivityReport } from './useModuleActivityReport';
