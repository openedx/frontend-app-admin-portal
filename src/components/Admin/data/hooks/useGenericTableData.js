import {
  useCallback, useMemo, useRef, useState,
} from 'react';
import { camelCaseObject } from '@edx/frontend-platform/utils';
import debounce from 'lodash.debounce';
import { logError } from '@edx/frontend-platform/logging';
import { sendEnterpriseTrackEvent } from '@edx/frontend-enterprise-utils';
import EVENT_NAMES from '../../../../eventTracking';

const applySortByToOptions = (sortBy, options, apiFieldsForColumnAccessor) => {
  if (!sortBy || sortBy.length === 0) {
    return;
  }
  const orderingStrings = sortBy.map(({ id, desc }) => {
    const apiFieldForColumnAccessor = apiFieldsForColumnAccessor[id];
    if (!apiFieldForColumnAccessor) {
      return undefined;
    }
    const apiFieldKey = apiFieldForColumnAccessor.key;
    return desc ? `-${apiFieldKey}` : apiFieldKey;
  }).filter(orderingString => !!orderingString);

  Object.assign(options, {
    ordering: orderingStrings.join(','),
  });
};

const useGenericTableData = (enterpriseId, tableId, fetchMethod, apiFields) => {
  const shouldTrackFetchEvents = useRef(false);
  const [isLoading, setIsLoading] = useState(true);
  const [tableData, setTableData] = useState({
    itemCount: 0,
    pageCount: 0,
    results: [],
  });

  const fetchTableData = useCallback(async (args) => {
    try {
      setIsLoading(true);
      const options = {
        page: args.pageIndex + 1,
        pageSize: args.pageSize,
      };
      applySortByToOptions(args.sortBy, options, apiFields);

      const response = await fetchMethod(enterpriseId, options);
      const data = camelCaseObject(response.data);

      setTableData({
        itemCount: data.count,
        pageCount: data.numPages ?? Math.floor(data.count / options.pageSize),
        results: data.results,
      });

      if (shouldTrackFetchEvents.current) {
        sendEnterpriseTrackEvent(
          enterpriseId,
          EVENT_NAMES.PROGRESS_REPORT.DATATABLE_SORT_BY_OR_FILTER,
          {
            tableId,
            ...options,
          },
        );
      } else {
        shouldTrackFetchEvents.current = true;
      }
    } catch (error) {
      logError(error);
    } finally {
      setIsLoading(false);
    }
  }, [enterpriseId, tableId, fetchMethod, apiFields]);

  const debouncedFetchTableData = useMemo(
    () => debounce(fetchTableData, 300),
    [fetchTableData],
  );

  return {
    isLoading,
    tableData,
    fetchTableData: debouncedFetchTableData,
  };
};

export default useGenericTableData;
