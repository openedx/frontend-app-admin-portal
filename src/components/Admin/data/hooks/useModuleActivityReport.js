import { useEffect, useState } from 'react';
import { logError } from '@edx/frontend-platform/logging';
import EnterpriseDataApiService from '../../../../data/services/EnterpriseDataApiService';

const useModuleActivityReport = ({
  enterpriseId, page, filters, searchQuery,
}) => {
  const [isLoading, setIsLoading] = useState(true);

  const [paginationData, setPaginationData] = useState({
    itemCount: 0,
    pageCount: 0,
    data: [],
  });

  useEffect(() => {
    // Reset the loading state
    setIsLoading(true);

    EnterpriseDataApiService.fetchEnterpriseModuleActivityReport(enterpriseId, {
      page: page + 1,
      search: searchQuery,
      ...filters,
    })
      .then((response) => {
        setPaginationData({
          itemCount: response.data.count,
          pageCount: response.data.num_pages,
          data: response.data.results,
          currentPage: response.data.currentPage,
        });

        // Reset the loading state
        setIsLoading(false);
      })
      .catch((err) => {
        logError(err);

        // Reset the loading state
        setIsLoading(false);
      });
  }, [
    enterpriseId,
    page,
    filters,
    searchQuery,
  ]);

  return {
    isLoading,
    paginationData,
  };
};

export default useModuleActivityReport;
