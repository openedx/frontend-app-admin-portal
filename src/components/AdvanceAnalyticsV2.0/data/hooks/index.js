import { useMemo } from 'react';
import { useQuery } from '@tanstack/react-query';

import { advanceAnalyticsQueryKeys } from '../constants';
import EnterpriseDataApiService from '../../../../data/services/EnterpriseDataApiService';

export { default as useEnterpriseEnrollmentsData } from './useEnterpriseEnrollmentsData';
export { default as useEnterpriseEngagementData } from './useEnterpriseEngagementData';
export { default as useEnterpriseCompletionsData } from './useEnterpriseCompletionsData';

export const useEnterpriseAnalyticsData = ({
  enterpriseCustomerUUID,
  key,
  startDate,
  endDate,
  granularity = undefined,
  calculation = undefined,
  groupUUID = undefined,
  currentPage = undefined,
  pageSize = undefined,
  queryOptions = {},
}) => {
  const requestOptions = {
    startDate, endDate, granularity, calculation, page: currentPage, pageSize, groupUUID,
  };
  return useQuery({
    queryKey: advanceAnalyticsQueryKeys[key](enterpriseCustomerUUID, requestOptions),
    queryFn: () => EnterpriseDataApiService.fetchAdminAnalyticsData(
      enterpriseCustomerUUID,
      key,
      requestOptions,
    ),
    staleTime: 0.5 * (1000 * 60 * 60), // 30 minutes. The time in milliseconds after data is considered stale.
    cacheTime: 0.75 * (1000 * 60 * 60), // 45 minutes. Cache data will be garbage collected after this duration.
    keepPreviousData: true,
    ...queryOptions,
  });
};

export const usePaginatedData = (data) => useMemo(() => {
  if (data) {
    return {
      data: data.results,
      pageCount: data.numPages,
      itemCount: data.count,
      currentPage: data.currentPage,
    };
  }

  return {
    itemCount: 0,
    pageCount: 0,
    data: [],
  };
}, [data]);

export const useEnterpriseAnalyticsAggregatesData = ({
  enterpriseCustomerUUID,
  startDate,
  endDate,
  queryOptions = {},
}) => {
  const requestOptions = {
    startDate, endDate,
  };
  return useQuery({
    queryKey: advanceAnalyticsQueryKeys.aggregates(enterpriseCustomerUUID, requestOptions),
    queryFn: () => EnterpriseDataApiService.fetchAdminAggregatesData(
      enterpriseCustomerUUID,
      requestOptions,
    ),
    staleTime: 0.5 * (1000 * 60 * 60), // 30 minutes. The time in milliseconds after data is considered stale.
    cacheTime: 0.75 * (1000 * 60 * 60), // 45 minutes. Cache data will be garbage collected after this duration.
    keepPreviousData: true,
    ...queryOptions,
  });
};
