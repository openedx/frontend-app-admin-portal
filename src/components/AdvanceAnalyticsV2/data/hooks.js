import { useMemo } from 'react';
import { useQuery } from '@tanstack/react-query';

import { advanceAnalyticsQueryKeys } from './constants';
import EnterpriseDataApiService from '../../../data/services/EnterpriseDataApiService';

export const useEnterpriseAnalyticsData = ({
  enterpriseCustomerUUID,
  key,
  startDate,
  endDate,
  granularity = undefined,
  calculation = undefined,
  currentPage = undefined,
  queryOptions = {},
}) => {
  const requestOptions = {
    startDate, endDate, granularity, calculation, page: currentPage,
  };
  return useQuery({
    queryKey: advanceAnalyticsQueryKeys[key](enterpriseCustomerUUID, requestOptions),
    queryFn: () => EnterpriseDataApiService.fetchAdminAnalyticsData(
      enterpriseCustomerUUID,
      key,
      requestOptions,
    ),
    staleTime: 0.5 * (1000 * 60 * 60), // 30 minutes. The time in milliseconds after data is considered stale.
    cacheTime: 0.75 * (1000 * 60 * 60), // 45 minutes. Cache data will be garbage collected after this duration.
    keepPreviousData: true,
    ...queryOptions,
  });
};

export const usePaginatedData = (data) => useMemo(() => {
  if (data) {
    return {
      data: data.results,
      pageCount: data.numPages,
      itemCount: data.count,
      currentPage: data.currentPage,
    };
  }

  return {
    itemCount: 0,
    pageCount: 0,
    data: [],
  };
}, [data]);
