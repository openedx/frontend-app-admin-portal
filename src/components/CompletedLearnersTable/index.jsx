import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import TableContainer from '../../containers/TableContainer';
import EnterpriseDataApiService from '../../data/services/EnterpriseDataApiService';

const CompletedLearnersTable = () => {
  const intl = useIntl();

  const tableColumns = [
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.completed.learners.table.user_email.column.heading',
        defaultMessage: 'Email',
        description: 'Column heading for the user email column in the completed learners table',
      }),
      key: 'user_email',
      columnSortable: true,
    },
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.completed.learned.table.completed_courses.column.heading',
        defaultMessage: 'Total Course Completed Count',
        description: 'Column heading for the completed courses column in the completed learners table',
      }),
      key: 'completed_courses',
      columnSortable: true,
    },
  ];

  const formatLearnerData = learners => learners.map(learner => ({
    ...learner,
    user_email: <span data-hj-suppress>{learner.user_email}</span>,
  }));

  return (
    <TableContainer
      id="completed-learners"
      className="completed-learners"
      fetchMethod={EnterpriseDataApiService.fetchCompletedLearners}
      columns={tableColumns}
      formatData={formatLearnerData}
      tableSortable
    />
  );
};

export default CompletedLearnersTable;
