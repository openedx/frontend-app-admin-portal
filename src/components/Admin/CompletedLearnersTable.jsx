import React, { useMemo } from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';
import { DataTable } from '@openedx/paragon';
import { useGenericTableData } from './data/hooks';
import EnterpriseDataApiService from '../../data/services/EnterpriseDataApiService';
import EmailCell from './EmailCell';

const CompletedLearnersTable = (enterpriseId) => {
  const intl = useIntl();
  const {
    isLoading,
    tableData,
    fetchTableData,
  } = useGenericTableData(
    enterpriseId,
    'completed-learners',
    EnterpriseDataApiService.fetchCompletedLearners,
    useMemo(() => ({
      userEmail: { key: 'user_email' },
      completedCourses: { key: 'completed_courses' },
    }), []),
  );

  return (
    <DataTable
      isSortable
      manualSortBy
      isPaginated
      manualPagination
      isLoading={isLoading}
      columns={[
        {
          Header: intl.formatMessage({
            id: 'admin.portal.lpr.completed.learners.table.user_email.column.heading',
            defaultMessage: 'Email',
            description: 'Column heading for the user email column in the completed learners table',
          }),
          accessor: 'userEmail',
          Cell: EmailCell,
        },
        {
          Header: intl.formatMessage({
            id: 'admin.portal.lpr.completed.learners.table.completed_courses.column.heading',
            defaultMessage: 'Total Course Completed Count',
            description: 'Column heading for the completed courses column in the completed learners table',
          }),
          accessor: 'completedCourses',
        },
      ]}
      initialState={{
        pageIndex: 20,
        pageSize: 0,
        sortBy: [{ id: 'completedCourses', desc: true }],
        selectedRowsOrdered: [],
      }}
      fetchData={fetchTableData}
      data={tableData.results}
      itemCount={tableData.itemCount}
      pageCount={tableData.pageCount}
    />
  );
};

export default CompletedLearnersTable;
