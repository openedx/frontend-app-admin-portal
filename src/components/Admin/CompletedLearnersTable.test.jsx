import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

import CompletedLearnersTable from './CompletedLearnersTable';
import { useGenericTableData } from './data/hooks';

const mockStore = configureMockStore([thunk]);
const enterpriseId = 'test-enterprise';

jest.mock('./data/hooks/useGenericTableData', () => (
  jest.fn().mockReturnValue({})
));

const store = mockStore({
  portalConfiguration: {
    enterpriseId,
  },
});

const CompletedLearnersWrapper = props => (
  <MemoryRouter>
    <IntlProvider locale="en">
      <Provider store={store}>
        <CompletedLearnersTable
          {...props}
        />
      </Provider>
    </IntlProvider>
  </MemoryRouter>
);

describe('CompletedLearnersTable', () => {
  it('renders empty state correctly', () => {
    useGenericTableData.mockReturnValue({
      isLoading: false,
      tableData: {
        results: [],
        itemCount: 0,
        pageCount: 0,
      },
      fetchTableData: jest.fn(),
    });
    const tree = renderer
      .create((
        <CompletedLearnersWrapper />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
