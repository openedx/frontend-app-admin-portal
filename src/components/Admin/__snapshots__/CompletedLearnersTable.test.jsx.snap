// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletedLearnersTable renders empty state correctly 1`] = `
<div
  className="pgn__data-table-layout-wrapper"
>
  <div
    className="pgn__data-table-layout-main"
  >
    <div
      className="pgn__data-table-wrapper"
    >
      <div
        className="pgn__data-table-status-bar"
        data-testid="table-control-bar"
      >
        <div
          className="pgn__data-table-status"
        >
          <div
            className="pgn__data-table-status-left"
          />
          <div
            className="pgn__data-table-actions-right"
          >
            <div
              className="pgn__data-table-toggle"
            />
            <div />
          </div>
        </div>
      </div>
      <div
        className="pgn__data-table-container"
        data-testid="data-table-container"
      >
        <table
          className="pgn__data-table is-striped"
          role="table"
        >
          <thead>
            <tr
              role="row"
            >
              <th
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Email
                  </span>
                  <span
                    className="pgn__icon"
                    data-testid="arrow-drop-up-down"
                    style={
                      {
                        "opacity": 0.5,
                      }
                    }
                  >
                    <svg
                      aria-hidden={true}
                      fill="none"
                      focusable={false}
                      height={24}
                      role="img"
                      viewBox="0 0 24 24"
                      width={24}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m7 10 5-5 5 5H7ZM7 14l5 5 5-5H7Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </span>
              </th>
              <th
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Total Course Completed Count
                  </span>
                  <span
                    className="pgn__icon"
                    data-testid="arrow-drop-down"
                  >
                    <svg
                      aria-hidden={true}
                      fill="none"
                      focusable={false}
                      height={24}
                      role="img"
                      viewBox="0 0 24 24"
                      width={24}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m7 10 5 5 5-5H7Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            role="rowgroup"
          />
        </table>
      </div>
      <div
        className="pgn__data-table-empty"
      >
        No results found
      </div>
      <div
        className="pgn__data-table-footer"
        data-testid="table-footer"
      >
        <nav
          aria-label="table pagination"
          className="pagination-minimal"
        >
          <div
            aria-atomic={true}
            aria-live="polite"
            aria-relevant="text"
            className="sr-only"
          >
            Page 21, Current Page, of -1
          </div>
          <ul
            className="pagination"
          >
            <li
              className="page-item"
            >
              <button
                aria-label="Previous, Page 20"
                className="btn-icon btn-icon-primary btn-icon-md previous page-link"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <span
                  className="btn-icon__icon-container"
                >
                  <span
                    className="pgn__icon btn-icon__icon"
                  >
                    <svg
                      aria-hidden={true}
                      fill="none"
                      focusable={false}
                      height={24}
                      role="img"
                      viewBox="0 0 24 24"
                      width={24}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M17.51 3.87 15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </li>
            <li
              className="page-item disabled"
            >
              <button
                aria-label="Next, Page 22"
                className="btn-icon btn-icon-primary btn-icon-md next page-link"
                disabled={true}
                onClick={[Function]}
                tabIndex="-1"
                type="button"
              >
                <span
                  className="btn-icon__icon-container"
                >
                  <span
                    className="pgn__icon btn-icon__icon"
                  >
                    <svg
                      aria-hidden={true}
                      fill="none"
                      focusable={false}
                      height={24}
                      role="img"
                      viewBox="0 0 24 24"
                      width={24}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m6.49 20.13 1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
